### PR TITLE
SELinux libs and tools: 2.4 -> 2.7

### PR DESCRIPTION
**Pull request by xeji**
###### Motivation for this change

Our SELinux 2.4 is >3 yrs old and produces some build fails with gcc7.
[WIP] since this update involves multiple packages which will be added here incrementally.

/cc @phreedom @andir

### CODE REVIEW

**Comment by Mic92** [comment]
Has anyone a working selinux on NixOS actually? I did not even manage to enable it.

**Comment by xeji** [comment]
I want to  give it a try, that's why I am doing this first.

**Comment by Mic92** [comment]
I would be glad to see this. I currently have to boot fedora to test selinux support in https://github.com/Mic92/cntr

**Comment by Mic92** [comment]
I would not be surprise that I was not able to enable it because of the age of selinux in comparison to the kernel.

**Comment by Mic92** [comment]
@GrahamcOfBorg build libsepool libselinux libsemanage policycoreutils checkpolicy

**Comment by xeji** [comment]
@GrahamcOfBorg build libsepol libselinux libsemanage policycoreutils checkpolicy

**Comment by xeji** [comment]
@GrahamcOfBorg build setools

**Comment by xeji** [comment]
@grahamcofborg build selinux-python

File: pkgs/os-specific/linux/setools/default.nix [hunk 1]
```diff
@@ -1,35 +1,42 @@
...
+  '';
 
-  NIX_CFLAGS_COMPILE = "-fstack-protector-all";
-  NIX_LDFLAGS = "-L${libsepol}/lib -L${libselinux}/lib";
+  setupPyBuildFlags = [ "-i" ];
 
-  nativeBuildInputs = [ autoreconfHook pkgconfig python swig2 bison flex ];
-  buildInputs = [ tcl libxml2 sqlite bzip2 ];
+  preBuild = ''
+    export SEPOL="${libsepol}/lib/libsepol.a"
```
Comment (Mic92) [inline pkgs/os-specific/linux/setools/default.nix]: Is `libsepol` supposed to be statically linked?

**Comment by xeji** [comment]
Yes, upstream code explicitly links libsepol.a  statically. Surprised me too but I see no reason to mess with that. It's not that big anyway.

**Comment by xeji** [comment]
@GrahamcOfBorg build selinux-sandbox

**Comment by xeji** [comment]
@grahamcofborg build libsepol libselinux libsemanage policycoreutils checkpolicy

**Comment by xeji** [comment]
These look good now. Not yet tested on an SELinux-enabled system though (I somehow doubt our current version ever was).

**Comment by xeji** [comment]
rebased to solve merge conflict

**Comment by dtzWill** [comment]
:+1:

**Comment by Mic92** [comment]
@xeji did you manage to get selinux running on nixos and if what flags where required?

**Comment by xeji** [comment]
@Mic92 not yet - it's on my list for the next few days. Testing on fedora would be good anyway to confirm the tools work independent of nixos setup.

**Comment by xeji** [comment]
@Mic92 some quick first tests show you need ` boot.kernelParams = [ "security=selinux" "selinux=1" ];` to boot nixos with selinux. Then the logs look promising: 
 
```
# dmesg | grep -i selinux
[    0.001588] SELinux:  Initializing.
[    0.001592] SELinux:  Starting in permissive mode
[    0.382732] SELinux:  Registering netfilter hooks
[    1.574033] systemd[1]: systemd 237 running in system mode. (+PAM +AUDIT -SELINUX +IMA +APPARMOR +SMACK -SYSVINIT +UTMP -LIBCRYPTSETUP +GCRYPT -GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID -ELFUTILS +KMOD +IDN2 -IDN -PCRE2 default-hierarchy=hybrid)
```
However, after boot selinux still shows as disabled: 
```
# sestatus
SELinux status:                 disabled
```
I suspect the `-SELINUX` in systemd build config prevents it from working, so we would probably have to override and rebuild systemd. Don't know how to do that (it's a separate NixOS repo ?). 
Anyway, this situation looks already better than selinux 2.4 which caused systemd-related kernel backtraces...

**Comment by dtzWill** [comment]
> I suspect the -SELINUX in systemd build config prevents it from working, so we would probably have to override and rebuild systemd. Don't know how to do that (it's a separate NixOS repo ?).
> Anyway, this situation looks already better than selinux 2.4 which caused systemd-related kernel backtraces...

I believe that means it wasn't available at compile-time (or that we forced it off with `-Dselinux=false`, which we don't)-- which seems like a good starting point.  No clue what else might need changing, if anything, though.

Also, yay!

**Comment by Mic92** [comment]
@xeji looking at systemd is a good hint. It is possible to enable selinux in selinux without recompiling everything by loading a custom systemd with `systemd.package =` in `configuration.nix`. I will skim over the code to see if I can find something interesting.

**Comment by xeji** [comment]
Looking at the systemd expression, we don't disable selinux. But we don't list `libselinux` (that's what I guess it would be) as build input either. So if we're lucky, just adding that may hopefully configure systemd with `+SELINUX`. I'll give that a try when I have time for a lengthy build...

**Comment by Mic92** [comment]
`selinux_init_load_policy` seems to enable selinux globally: 
https://github.com/systemd/systemd/blob/master/src/core/selinux-setup.c#L81

**Comment by xeji** [comment]
https://github.com/systemd/systemd/blob/ffb3c2bd705409ca5bbbb9ccef4c59349ea787cf/meson.build#L850 
if libselinux is present, systemd builds with it

**Comment by xeji** [comment]
The last commit adds a `withSelinux` option to systemd to build it with selinux support, `false` by default so this should not trigger a rebuild. Now  nixos boots with selinux enabled using [this config](https://gist.github.com/xeji/78d4abbb493fc5a5b6f069b41ebdafbe).
For a fully functional selinux-enabled system we might also have to rebuild other packages like `coreutils` and `util-linux`  with selinux support according to the [Arch wiki](https://wiki.archlinux.org/index.php/SELinux).

**Comment by Mic92** [comment]
@xeji I suppose this selinux snippet could become a module later.

**Comment by pbogdan** [comment]
Just a heads up - it seems `libsepol` doesn't build on i686-linux (https://hydra.nixos.org/build/72006419) which is holding up the build of the minimal iso for that platform (part of the unstable channel tested job). No idea how to fix it though :-(.

**Comment by xeji** [comment]
@pbogdan thanks. The error looks like it can be fixed by ignoring a warning, I'll give that a try later, cannot hurt. But I have no idea how to test a i686 build. Is there a way to cross-build this on a x86_64 machine? Never tried...

**Comment by pbogdan** [comment]
Yeah, at least for i686-linux you can build it with `nix-build -A libsepol --argstr system i686-linux`.

**Comment by xeji** [comment]
@Mic92 a selinux module would be nice, let's keep that in mind. Probably quite a bit of work since we need some sort of policy builder/patcher tool to create policies that fit a nixos system.

**Comment by Mic92** [comment]
@xeji I think supporting containers would be already useful. docker/lxc/systemd-nspawn already apply profiles.

File: pkgs/os-specific/linux/selinux-sandbox/default.nix [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  nativeBuildInputs = [ wrapPython ];
+  buildInputs = [ bash coreutils libcap_ng policycoreutils python3 xorgserver openbox xmodmap dbus ];
+  propagatedBuildInputs = [ pygobject3 selinux-python ];
+
+  postPatch = ''
+    # Fix setuid install
+    substituteInPlace Makefile --replace "-m 4755" "-m 755"
+    substituteInPlace sandboxX.sh \
+      --replace "#!/bin/sh" "#!${bash}/bin/sh" \
+      --replace "/usr/share/sandbox/start" "${placeholder "out"}/share/sandbox/start" \
```
Comment (Infinisil) [inline pkgs/os-specific/linux/selinux-sandbox/default.nix]: This doesn't work for Nix 1.11 users, please correct this as seen in https://github.com/NixOS/nixpkgs/pull/37860

**Comment by xeji** [comment]
@Infinisil why does it matter? That placeholder is in a new package so this shouldn't break anything nix 1.11 users had before, or does it?

**Comment by Infinisil** [comment]
The minimum version required is still 1.11, not 2.0: https://github.com/NixOS/nixpkgs/blob/master/lib/minver.nix

**Comment by jtojnar** [comment]
Fixed in 025881c2364

**Comment by xeji** [comment]
Thanks @jtojnar !

**Pull request merged**